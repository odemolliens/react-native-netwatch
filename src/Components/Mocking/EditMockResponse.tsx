import { TextInput } from 'react-native-paper';
import { Button, KeyboardAvoidingView, Platform, ScrollView, StyleSheet, Switch, Text, View } from 'react-native';
import * as React from 'react';
import { useCallback, useContext, useEffect } from 'react';
import { ThemeContext } from '../../Theme';
import { NavBar } from '../NavBar';
import { extractURL, mockRequestWithResponse, MockResponse } from './utils';

export function EditMockResponse(props: { mockResponse: MockResponse; onPressBack: () => void; update: boolean }) {
  const theme = useContext(ThemeContext);

  const [url, setUrl] = React.useState<string>();
  const [method, setMethod] = React.useState<string>(); // [GET, POST, PUT, PATCH, DELETE
  const [headers, setHeaders] = React.useState<string>();
  const [body, setBody] = React.useState<string>();
  const [code, setCode] = React.useState<string>();
  const [active, setActive] = React.useState<boolean>(true);
  const [to, setTo] = React.useState<string>();

  useEffect(() => {
    setUrl(props.mockResponse.url);
    setMethod(props.mockResponse.method);
    setHeaders(props.mockResponse.headers);
    setBody(props.mockResponse.response);
    setCode(props.mockResponse.statusCode?.toString());
    setTo(props.mockResponse.timeout?.toString());
    setActive(props.update ? !!props.mockResponse.active : true);
  }, [
    props.mockResponse.headers,
    props.mockResponse.method,
    props.mockResponse.response,
    props.mockResponse.statusCode,
    props.mockResponse.timeout,
    props.mockResponse.url,
    props.mockResponse.active,
    props.update,
  ]);

  const saveMockResponse = useCallback(() => {
    if (url) {
      mockRequestWithResponse({
        url: extractURL(url),
        method,
        headers,
        response: body,
        statusCode: code ? parseInt(code, 10) : 0,
        timeout: to ? parseInt(to, 10) : 0,
        active,
        date: Date.now(),
      });
    }
  }, [body, code, headers, method, to, url, active]);

  return (
    <KeyboardAvoidingView style={styles.container} behavior={Platform.OS === 'ios' ? 'padding' : 'height'}>
      <NavBar
        title={'Edit Response Mock'}
        onPressBack={props.onPressBack}
        rightComponent={<Button title={'Save'} onPress={saveMockResponse} />}
      />

      <ScrollView
        keyboardDismissMode={'on-drag'}
        style={{ backgroundColor: theme.secondaryColor }}
        contentContainerStyle={styles.scrollView}
      >
        <View style={{ flex: 1, width: '100%' }}>
          <View style={{ margin: 15 }}>
            <Text style={{ color: theme.textColorOne }}>Leave blank entries to keep unchanged.</Text>
          </View>
          <View style={{ flexDirection: 'row', margin: 15 }}>
            <Text style={{ flex: 1, color: theme.textColorOne, fontSize: 14, marginBottom: 10 }}>
              Mock response active
            </Text>
            <Switch value={active} onValueChange={setActive} />
          </View>

          <TextInput mode={'flat'} style={styles.textInput} label={'URL'} onChangeText={setUrl} value={url} />
          <TextInput
            mode={'flat'}
            keyboardType={'numeric'}
            style={styles.textInput}
            label={'Response Code'}
            onChangeText={setCode}
            value={code}
          />
          <TextInput
            mode={'flat'}
            keyboardType={'numeric'}
            style={styles.textInput}
            label={'Timeout value'}
            onChangeText={setTo}
            value={to}
          />
          <TextInput
            mode={'flat'}
            style={styles.textInput}
            label={'Response Method'}
            onChangeText={setMethod}
            value={method}
          />
          <TextInput
            mode={'flat'}
            style={[styles.textInput, styles.multiLine]}
            value={headers}
            label="Headers"
            multiline
            onChangeText={setHeaders}
          />
          <TextInput
            mode={'flat'}
            style={[styles.textInput, styles.multiLine]}
            label="Body"
            multiline
            onChangeText={setBody}
            value={body}
          />
        </View>
      </ScrollView>
    </KeyboardAvoidingView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  subheading: {
    paddingHorizontal: 16,
    paddingVertical: 8,
    marginBottom: 10,
  },
  button: {
    justifyContent: 'center',
    alignItems: 'center',
    height: 56,
    width: 56,
    borderLeftWidth: 1,
  },
  scrollView: {
    paddingTop: 26,
    paddingBottom: 20,
  },
  textInput: {
    margin: 10,
  },
  multiLine: {
    height: 200,
  },
});
