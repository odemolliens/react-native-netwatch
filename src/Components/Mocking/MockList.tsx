import { Button, ScrollView, StyleSheet, Text, TouchableOpacity, View } from 'react-native';
import React, { useContext, useEffect, useState } from 'react';
import { ThemeContext } from '../../Theme';
import { NavBar } from '../NavBar';
import { Divider } from 'react-native-paper';
import { clearMockResponses, FILE_PATH, getMockResponses, MockResponse, resetMockResponses } from './utils';
import FeatherIcon from 'react-native-vector-icons/Feather';
import Share from 'react-native-share';
import RNFS from 'react-native-fs';
import Clipboard from '@react-native-clipboard/clipboard';

export function NavbarRightSide(props: {
  mockResponsesCopy: MockResponse[];
  setMockResponsesCopy: (mockResponsesCopy: MockResponse[]) => void;
}) {
  const theme = useContext(ThemeContext);

  return (
    <>
      <TouchableOpacity
        onPress={() => {
          Clipboard.getString().then(responses => {
            if (responses) {
              resetMockResponses(responses);
              props.setMockResponsesCopy(getMockResponses());
            }
          });
        }}
        style={[styles.button, { borderLeftWidth: 0 }]}
      >
        <FeatherIcon name="download" color={theme.textColorOne} size={24} />
      </TouchableOpacity>
      {props.mockResponsesCopy.length > 0 && (
        <TouchableOpacity
          style={[styles.button, { borderLeftWidth: 0 }]}
          onPress={() => {
            RNFS.copyFile(FILE_PATH, `${RNFS.DocumentDirectoryPath}/mocks-preset.json`)
              .then(() => {
                Share.open({
                  title: 'Export Mock preset',
                  url: `${RNFS.DocumentDirectoryPath}/mocks-preset.json`,
                  type: 'text/plain',
                  // excludedActivityTypes: []
                }).then(() => {
                  RNFS.unlink(`${RNFS.DocumentDirectoryPath}/mocks-preset.json`);
                });
              })
              .catch(console.error);
          }}
        >
          <FeatherIcon name="upload" color={theme.textColorOne} size={24} />
        </TouchableOpacity>
      )}
      <Button
        title={'Clear'}
        onPress={() => {
          clearMockResponses();
          props.setMockResponsesCopy([]);
        }}
      />
    </>
  );
}

export function MockList(props: {
  showEdit: (mockResponse: MockResponse, update: boolean) => void;
  onPressBack: () => void;
}) {
  const theme = useContext(ThemeContext);

  const [mockResponsesCopy, setMockResponsesCopy] = useState<MockResponse[]>([]);

  useEffect(() => {
    setMockResponsesCopy(getMockResponses());
  }, []);

  return (
    <View style={styles.container}>
      <NavBar
        icon="x"
        title={'Mock List'}
        onPressBack={props.onPressBack}
        rightComponent={
          <NavbarRightSide mockResponsesCopy={mockResponsesCopy} setMockResponsesCopy={setMockResponsesCopy} />
        }
      />

      <ScrollView style={{ backgroundColor: theme.secondaryColor }} contentContainerStyle={styles.scrollView}>
        {mockResponsesCopy.length === 0 && (
          <Text style={{ color: theme.textColorOne, alignSelf: 'center' }}>No mock created yet</Text>
        )}
        {mockResponsesCopy.length > 0 && (
          <>
            <View style={{ flexDirection: 'row' }}>
              <Text style={[{ flex: 1, color: theme.textColorOne }, styles.title]}>URL</Text>
              <Text style={[styles.title, { color: theme.textColorOne }]}>Method</Text>
              <Text style={[styles.title, { color: theme.textColorOne }]} />
            </View>

            {mockResponsesCopy.map((mock, index) => (
              <TouchableOpacity key={`mock-${index}`} onPress={() => props.showEdit(mock, true)}>
                <View style={{ flexDirection: 'row' }}>
                  <Text style={[{ flex: 1, color: mock.active ? theme.primaryColor : 'grey' }, styles.cell]}>
                    {mock.url}
                  </Text>
                  <Text style={[styles.cell, { color: theme.textColorOne }]}>{mock.method}</Text>
                  <View style={[styles.cell, { minWidth: 30 }]}>
                    <FeatherIcon name="arrow-right" color={theme.textColorOne} size={24} />
                  </View>
                </View>
                <Divider />
              </TouchableOpacity>
            ))}
          </>
        )}
      </ScrollView>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    height: '100%',
    width: '100%',
    top: 0,
    left: 0,
  },
  subheading: {
    paddingHorizontal: 16,
    paddingVertical: 8,
    marginBottom: 10,
  },
  button: {
    justifyContent: 'center',
    alignItems: 'center',
    height: 56,
    width: 56,
    borderLeftWidth: 1,
  },
  scrollView: {
    paddingTop: 26,
    paddingBottom: 20,
  },

  title: {
    fontWeight: 'bold',
    margin: 10,
    minWidth: 50,
  },

  cell: {
    margin: 10,
    alignSelf: 'flex-start',
    minWidth: 50,
  },
});
