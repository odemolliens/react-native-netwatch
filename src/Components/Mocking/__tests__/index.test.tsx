import * as React from 'react';
import { MockingNavigator } from '../index';
import { shallow, ShallowWrapper } from 'enzyme';
import { MockList } from '../MockList';
import { useState } from 'react';
import { EditMockResponse } from '../EditMockResponse';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
  useEffect: (f: () => void) => f(),
}));

describe('Test Mock navigator Screen', () => {
  let component: ShallowWrapper;
  const onPressBack = jest.fn();
  const setShowList = jest.fn();
  const setShowEdit = jest.fn();
  const setUpdate = jest.fn();
  const setMockResponse = jest.fn();
  const mockResponse = {
    url: 'https://www.google.com',
    method: 'GET',
    headers: '',
    response: '',
    statusCode: 0,
    timeout: 0,
    active: true,
    date: 0,
  };

  it('should render mock list properly', () => {
    givenShowMockList();
    component = shallow(<MockingNavigator onPressBack={() => jest.fn()} update={false} />);
    expect(component).toMatchSnapshot(component);
  });

  it('should render Edit mock screen properly', () => {
    givenShowEditMockResponse();
    component = shallow(<MockingNavigator mockResponse={mockResponse} onPressBack={onPressBack} update={false} />);
    expect(component).toMatchSnapshot(component);
  });

  it('should close Mock navigator when pressing cross button', () => {
    givenShowMockList();
    component = shallow(<MockingNavigator onPressBack={onPressBack} update={false} />);
    component.find(MockList).at(0).simulate('pressBack');
    expect(onPressBack).toHaveBeenCalled();
  });

  it('should show Edit Mock Response screen', function () {
    givenShowMockList();
    component = shallow(<MockingNavigator onPressBack={() => jest.fn()} update={false} />);
    component.find(MockList).at(0).props().showEdit(mockResponse, true);
    expect(setShowList).toHaveBeenCalledWith(false);
    expect(setShowEdit).toHaveBeenCalledWith(true);
    expect(setMockResponse).toHaveBeenCalledWith(mockResponse);
    expect(setUpdate).toHaveBeenCalledWith(true);
  });

  it('should show List screen when closing cross button on edit mock response screen', function () {
    givenShowEditMockResponse();
    component = shallow(<MockingNavigator onPressBack={onPressBack} update={false} />);
    component.find(EditMockResponse).simulate('pressBack');
    expect(setShowList).toHaveBeenCalledWith(true);
    expect(setShowEdit).toHaveBeenCalledWith(false);
  });

  function givenShowMockList() {
    (useState as jest.Mock).mockReturnValueOnce([true, setShowList]);
    (useState as jest.Mock).mockReturnValueOnce([false, setShowEdit]);
    (useState as jest.Mock).mockReturnValueOnce([mockResponse, setMockResponse]);
    (useState as jest.Mock).mockReturnValueOnce([false, setUpdate]);
  }

  function givenShowEditMockResponse() {
    (useState as jest.Mock).mockReturnValueOnce([false, setShowList]);
    (useState as jest.Mock).mockReturnValueOnce([true, setShowEdit]);
    (useState as jest.Mock).mockReturnValueOnce([mockResponse, setMockResponse]);
    (useState as jest.Mock).mockReturnValueOnce([false, setUpdate]);
  }
});
