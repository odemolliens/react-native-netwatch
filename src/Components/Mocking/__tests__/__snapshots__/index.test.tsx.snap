// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Mock navigator Screen should render Edit mock screen properly 1`] = `Object {}`;

exports[`Test Mock navigator Screen should render mock list properly 1`] = `Object {}`;
